README.md
# Dash Iconify

Dash Iconify based on Iconify is a Dash component library which brings over 100,000 vector icons.

## Table of contents

- [Installation](#installation)
- [Quickstart](#quickstart)
- [Using with dmc](#using-with-dmc)
- [Keyword Arguments](#keyword-arguments)

## Installation

```bash
pip install dash-iconify
```

## Quickstart

```python
from dash_iconify import DashIconify
from dash import Dash

app = Dash(__name__)

app.layout = DashIconify(
    icon="ion:logo-github",
    width=30,
    height=30,
    rotate=1,
    flip="horizontal",
)

if __name__ == "__main__":
    app.run_server(debug=True)

```

## Using with dmc

[Dash Mantine Components](https://github.com/snehilvj/dash-mantine-components) enables using icons natively.

```python
import dash_mantine_components as dmc
from dash_iconify import DashIconify

button = dmc.Button("Send Mail", leftIcon=[
    DashIconify(icon="fluent:folder-mail-16-filled")
])
```

## Keyword Arguments

Visit this site to browse all the available icons: <https://icon-sets.iconify.design/>

```git
Keyword arguments:

- id (string; optional):
    The ID used to identify this component in Dash callbacks.

- color (string; optional):
    Color.

- flip (a value equal to: "horizontal", "vertical"; optional):
    Flip the icon horizontally or vertically.

- height (number; optional):
    Icon height.

- icon (string; optional):
    Icon name is a string, which has 3 parts: @api-provider : icon-prefix : icon-name
    provider points to API source. Starts with "@", can be empty (empty value is used for public Iconify API).
    prefix is name of icon set.
    name is name of icon.

- inline (boolean; optional):
    Toggles inline or block mode.

- rotate (a value equal to: 0, 1, 2, 3; optional):
    Rotates icon, 0: 0 deg, 1: 90 deg, 2: 180 deg, 3: 270 deg.

- style (dict; optional):
    Inline style.

- width (number; optional):
    Icon width.
```


CHANGELOG.md
## 0.1.2

Published dash_iconify to npm.

### Added

- `className` prop.

## 0.0.1 

### Added
 - First release of Iconify for Plotly Dash.

_validate_init.py
"""
DO NOT MODIFY
This file is used to validate your publish settings.
"""
from __future__ import print_function

import os
import sys
import importlib


components_package = 'dash_iconify'

components_lib = importlib.import_module(components_package)

missing_dist_msg = 'Warning {} was not found in `{}.__init__.{}`!!!'
missing_manifest_msg = '''
Warning {} was not found in `MANIFEST.in`!
It will not be included in the build!
'''

with open('MANIFEST.in', 'r') as f:
    manifest = f.read()


def check_dist(dist, filename):
    # Support the dev bundle.
    if filename.endswith('dev.js'):
        return True

    return any(
        filename in x
        for d in dist
        for x in (
            [d.get('relative_package_path')]
            if not isinstance(d.get('relative_package_path'), list)
            else d.get('relative_package_path')
        )
    )


def check_manifest(filename):
    return filename in manifest


def check_file(dist, filename):
    if not check_dist(dist, filename):
        print(
            missing_dist_msg.format(filename, components_package, '_js_dist'),
            file=sys.stderr
        )
    if not check_manifest(filename):
        print(missing_manifest_msg.format(filename),
              file=sys.stderr)


for cur, _, files in os.walk(components_package):
    for f in files:

        if f.endswith('js'):
            # noinspection PyProtectedMember
            check_file(components_lib._js_dist, f)
        elif f.endswith('css'):
            # noinspection PyProtectedMember
            check_file(components_lib._css_dist, f)
        elif not f.endswith('py'):
            check_manifest(f)


webpack.config.js
const path = require('path');
const TerserPlugin = require('terser-webpack-plugin');
const webpack = require('webpack');
const WebpackDashDynamicImport = require('@plotly/webpack-dash-dynamic-import');
const packagejson = require('./package.json');

const dashLibraryName = packagejson.name.replace(/-/g, '_');

module.exports = (env, argv) => {

    let mode;

    const overrides = module.exports || {};

    // if user specified mode flag take that value
    if (argv && argv.mode) {
        mode = argv.mode;
    }

    // else if configuration object is already set (module.exports) use that value
    else if (overrides.mode) {
        mode = overrides.mode;
    }

    // else take webpack default (production)
    else {
        mode = 'production';
    }

    let filename = (overrides.output || {}).filename;
    if(!filename) {
        const modeSuffix = mode === 'development' ? 'dev' : 'min';
        filename = `${dashLibraryName}.${modeSuffix}.js`;
    }

    const entry = overrides.entry || {main: './src/lib/index.js'};

    const devtool = overrides.devtool || 'source-map';

    const externals = ('externals' in overrides) ? overrides.externals : ({
        react: 'React',
        'react-dom': 'ReactDOM',
        'plotly.js': 'Plotly',
        'prop-types': 'PropTypes',
    });

    return {
        mode,
        entry,
        output: {
            path: path.resolve(__dirname, dashLibraryName),
            chunkFilename: '[name].js',
            filename,
            library: dashLibraryName,
            libraryTarget: 'window',
        },
        devtool,
        externals,
        module: {
            rules: [
                {
                    test: /\.jsx?$/,
                    exclude: /node_modules/,
                    use: {
                        loader: 'babel-loader',
                    },
                },
                {
                    test: /\.css$/,
                    use: [
                        {
                            loader: 'style-loader',
                            options: {
                                insertAt: 'top'
                            }
                        },
                        {
                            loader: 'css-loader',
                        },
                    ],
                },
            ],
        },
        optimization: {
            minimizer: [
                new TerserPlugin({
                    sourceMap: true,
                    parallel: true,
                    cache: './.build_cache/terser',
                    terserOptions: {
                        warnings: false,
                        ie8: false
                    }
                })
            ],
            splitChunks: {
                name: true,
                cacheGroups: {
                    async: {
                        chunks: 'async',
                        minSize: 0,
                        name(module, chunks, cacheGroupKey) {
                            return `${cacheGroupKey}-${chunks[0].name}`;
                        }
                    },
                    shared: {
                        chunks: 'all',
                        minSize: 0,
                        minChunks: 2,
                        name: 'dash_iconify-shared'
                    }
                }
            }
        },
        plugins: [
            new WebpackDashDynamicImport(),
            new webpack.SourceMapDevToolPlugin({
                filename: '[file].map',
                exclude: ['async-plotlyjs']
            })
        ]
    }
};


dash_iconify/__init__.py
from __future__ import print_function as _

import os as _os
import sys as _sys
import json

import dash as _dash

# noinspection PyUnresolvedReferences
from ._imports_ import *
from ._imports_ import __all__

if not hasattr(_dash, '__plotly_dash') and not hasattr(_dash, 'development'):
    print('Dash was not successfully imported. '
          'Make sure you don\'t have a file '
          'named \n"dash.py" in your current directory.', file=_sys.stderr)
    _sys.exit(1)

_basepath = _os.path.dirname(__file__)
_filepath = _os.path.abspath(_os.path.join(_basepath, 'package-info.json'))
with open(_filepath) as f:
    package = json.load(f)

package_name = package['name'].replace(' ', '_').replace('-', '_')
__version__ = package['version']

_current_path = _os.path.dirname(_os.path.abspath(__file__))

_this_module = _sys.modules[__name__]

async_resources = ["DashIconify",]

_js_dist = []

_js_dist.extend(
    [
        {
            "relative_package_path": "async-{}.js".format(async_resource),
            "external_url": (
                "https://unpkg.com/{0}@{2}"
                "/{1}/async-{3}.js"
            ).format(package_name, __name__, __version__, async_resource),
            "namespace": package_name,
            "async": True,
        }
        for async_resource in async_resources
    ]
)

# TODO: Figure out if unpkg link works
_js_dist.extend(
    [
        {
            "relative_package_path": "async-{}.js.map".format(async_resource),
            "external_url": (
                "https://unpkg.com/{0}@{2}"
                "/{1}/async-{3}.js.map"
            ).format(package_name, __name__, __version__, async_resource),
            "namespace": package_name,
            "dynamic": True,
        }
        for async_resource in async_resources
    ]
)

_js_dist.extend(
    [
        {
            'relative_package_path': 'dash_iconify.min.js',
    'external_url': 'https://unpkg.com/{0}@{2}/{1}/{1}.min.js'.format(
                package_name, __name__, __version__),
            'namespace': package_name
        },
        {
            'relative_package_path': 'dash_iconify.min.js.map',
    'external_url': 'https://unpkg.com/{0}@{2}/{1}/{1}.min.js.map'.format(
                package_name, __name__, __version__),
            'namespace': package_name,
            'dynamic': True
        }
    ]
)

_css_dist = []


for _component in __all__:
    setattr(locals()[_component], '_js_dist', _js_dist)
    setattr(locals()[_component], '_css_dist', _css_dist)


src/lib/components/DashIconify.react.js
import React, { Component } from "react";
import PropTypes from "prop-types";
import { DashIconify as RealComponent } from "../LazyLoader";

/**
 * Iconify for Plotly Dash. Visit this site to browse all the available icons: https://icon-sets.iconify.design/
 */
export default class DashIconify extends Component {
    render() {
        return (
            <React.Suspense fallback={null}>
                <RealComponent {...this.props} />
            </React.Suspense>
        );
    }
}

DashIconify.defaultProps = {};

DashIconify.propTypes = {
    /**
     * Often used with CSS to style elements with common properties
     */
    className: PropTypes.string,

    /**
     * Color
     */
    color: PropTypes.string,

    /**
     * Flip the icon horizontally or vertically
     */
    flip: PropTypes.oneOf(["horizontal", "vertical"]),

    /**
     * Icon height
     */
    height: PropTypes.number,

    /**
     * Icon name is a string, which has 3 parts: @api-provider : icon-prefix : icon-name
     * provider points to API source. Starts with "@", can be empty (empty value is used for public Iconify API).
     * prefix is name of icon set.
     * name is name of icon.
     */
    icon: PropTypes.string.isRequired,

    /**
     * The ID used to identify this component in Dash callbacks.
     */
    id: PropTypes.string,

    /**
     * Toggles inline or block mode.
     */
    inline: PropTypes.bool,

    /**
     * Rotates icon, 0: 0 deg, 1: 90 deg, 2: 180 deg, 3: 270 deg
     */
    rotate: PropTypes.oneOf([0, 1, 2, 3]),

    /**
     * Dash-assigned callback that should be called to report property changes
     * to Dash, to make them available for callbacks.
     */
    setProps: PropTypes.func,

    /**
     * Inline style
     */
    style: PropTypes.object,

    /**
     * Icon width
     */
    width: PropTypes.number,
};

export const defaultProps = DashIconify.defaultProps;
export const propTypes = DashIconify.propTypes;
